File: good bye 2023/e.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

unordered_map<int, vector<int>> g;
unordered_map<int, int> val;
unordered_map<int, int> f;
unordered_map<int, int> c;

int dfs(int node) {
    c[val[node]]++;
    f[node] = c.size();

    for (int ch : g[node])
        dfs(ch);

    c[val[node]]--;
    if (c[val[node]] == 0)
        c.erase(val[node]);

    return f[node];
}

int dfs1(int node) {
    if (g[node].size() == 2) {
        int left = dfs1(g[node][0]);
        int right = dfs1(g[node][1]);
        int maxm = max((left - f[node] + 1) * (right - f[node] + 1), 1);
        return max({ left, right, f[node] });
    } else if (g[node].size() == 1) {
        int left = dfs1(g[node][0]);
        return max(left, f[node]);
    } else if (g[node].size() == 0) {
        return f[node];
    } else {
        int one = dfs1(g[node][0]);
        int two = dfs1(g[node][1]);
        for (int i = 2; i < g[node].size(); ++i) {
            int ch = g[node][i];
            if (two < dfs1(ch) && dfs1(ch) < one)
                two = dfs1(ch);
            if (dfs1(ch) >= one) {
                int t = one;
                one = dfs1(ch);
                two = t;
            }
        }
        return max((one - f[node] + 1) * (two - f[node] + 1), 1);
    }
}

int main() {
    int T;
    cin >> T;

    while (T--) {
        int n;
        cin >> n;
        g.clear();
        val.clear();
        f.clear();
        c.clear();

        vector<int> par(n);
        for (int i = 1; i < n; ++i) {
            int parent;
            cin >> parent;
            par[i] = parent;
            g[parent].push_back(i + 1);
        }

        for (int i = 1; i <= n; ++i) {
            int v;
            cin >> v;
            val[i] = v;
        }

        dfs(1);
        cout << dfs1(1) << endl;
    }

    return 0;
}
d.

File: good bye 2023/e.py
#may give tle due to recursive approach
from collections import defaultdict
for _ in range(int(input())):
    n=int(input())
    par=list(map(int,input().split()))   #par[node-1]
    val=list(map(int,input().split()))   #val[node-1]
    g=defaultdict(list)
    for i in range(n-1):g[par[i]].append(i+2)


    f={}
    c=defaultdict(int)
    def dfs(node):
        c[val[node-1]]+=1
        f[node]=len(c)

        for ch in g[node]:dfs(ch)

        c[val[node-1]]-=1
        if c[val[node-1]]==0:del c[val[node-1]]
        

    dfs(1)

    def dfs1(node):

        if len(g[node])==2:
            left=dfs1(g[node][0])
            right=dfs1(g[node][1])
            maxm[0]=max(maxm[0],(left-f[node]+1)*(right-f[node]+1))
            return max(left,right,f[node])
        
        elif len(g[node])==1:
            left=dfs1(g[node][0])
            maxm[0]=max(maxm[0],left-f[node]+1)
            return max(left,f[node])
        elif len(g[node])==0:
            maxm[0]=max(maxm[0],1)
            return f[node]
        else:
            one=max(dfs1(g[node][0]),dfs1(g[node][1]))
            two=min(dfs1(g[node][0]),dfs1(g[node][1]))
            for i in range(2,len(g[node])):
                ch=g[node][i]
                if two<dfs1(ch)<one:two=dfs1(ch)
                if dfs1(ch)>=one:
                    t=one
                    one=dfs1(ch)
                    two=t 
            maxm[0]=max(maxm[0],(one-f[node]+1)*(two-f[node]+1))
            return max(one,f[node])

    maxm=[0]
    dfs1(1)
    print(maxm[0])

